README.md
# mail-map
PHP Library for Imap Connections

# Requirements
* >= php 5.6
* php imap extension
* php mb_string extension

# Installation
```
composer require mail-map/mail-map
```

# Usage
This is a small client library that interfaces with php's imap library.

It is primarily intended for reading emails from a mail server, and does not currently support anything much more complex than that. The one caveat to that is that there is currently support for moving messages between mailboxes.

This was designed to be highly configurable and extendable. All the primary classes in the library are injected into one another via contract implementation, so its very easy to extend the classes or create your own implementations in order to hook into the library at any part of the process of making imap connection to returning the found emails.

A simple class that will query imap connections has been provided, however you are encouraged to extend this, or write your own in order to get more complex results.

## Example Usage
```php
// Instantiate a new MailMap with configuration for logging into the mail server
$mailMap = new \MailMap\MailMap([
    'host' => 'imap.gmail.com',     // Required argument
    'user' => 'example@gmail.com',  // Required argument
    'password' => 'password',       // Required argument
    'port' => 993,                  // Optional - Default 993
    'service' => 'imap',            // Optional - Default 'imap'
    'encryption' => 'ssl'           // Optional - Default 'ssl'
]);

// The 'query' method takes the name of the mailbox to query, and a 'callable'
// to add query conditions. The callable will receive a query object, and must
// return a query object
$emails = $mailMap->query('My-Inbox', function ($query) {
    return $query->since('07-Sep-2016')->limit(3);
});
```

## MailMap\\MailMap Api
A helper class for creating imap connections and searching emails

### query($inbox = 'INBOX', callable $queryCall = null)
Execute a query on the connection
* string $inbox - Name of the inbox to query on
* callable $queryCall - Callable that takes a MailMap\\Query Object as an argument. Conditions, sorting, and limits can be set on the query in the callable on the query object. The callable must return the query object.

Returns array of `MailMap\Mail` objects.

### mailboxes($mailboxSearch = '\*', $withConnection = false)
Get a list of available mailboxes on the connection
* string $mailboxSearch - Regex to search for mailboxes
* bool $withConnection - Mailboxes are returned from the imap library with the connection string as a prefix. By default, this will be stripped off, but passing in `true` will leave the connection string.

Returns array of strings.

## MailMap\\Query Object Api
Object that is passed into the `MailMap\MailMap`'s `query` method. This object contains methods that will add conditions, sorting, and limiting to the executed query.

### Conditions
Methods for adding conditions correspond to the [imap_search criteria in the php docs pages](http://php.net/manual/en/function.imap-search.php). There is an available method for each of these criteria.

The condition methods are as follows
* all()
* answered()
* bcc(string $match)
* before(string $date)
* body(string $match)
* cc(string $match)
* deleted()
* flagged()
* from(string $match)
* keyword(string $match)
* new()
* old()
* on(string $date)
* recent()
* seen()
* since(string $date)
* subject(string $match)
* text(string $match)
* to($match)
* unanswered()
* undeleted()
* unflagged()
* unkeyword()
* unseen()

See the [php docs on searching](http://php.net/manual/en/function.imap-search.php) for more details and on how to format arguments for these methods.

Example usage - Finding all deleted emails from `example@gmail.com` on August 1st 2015
```php
$query->deleted()->from('example@gmail.com')->on('01-Aug-2015');
```

### Sorting
The php imap library only supports one sort on each query, so additional sorting must be done elsewhere. Sort order can be added to the query by calling the 'order' method, and providing the order flag, and the direction to sort (default is 'desc')

Example usage - Sort the messages by subject in ascending order
```php
$query->order(SORTSUBJECT, 'asc');
```

Details on sorting, and the available sort flags can be found on the [imap_sort criteria in the php docs](http://php.net/manual/en/function.imap-sort.php).

### Limiting
The php imap library does not support limiting results. For this reason, there is a default maximum limit set on the returned results for queries of 100 as a safety net. This can be overwritten by simply passing in a larger value into the limit method, or by changing the static property on the Query class.

Change globally
```php
MailMap\Query::$max = 200;
```

Or set inline on an individual query
```php
$query->limit(200);
```

## MailMap\\Map Api
Objects that are returned by the `MailMap\MailMap`'s `query` method. The objects are generated by the `MailMap\MailFactory` class. As stated above, the factory and mail object classes can easily be extended, or a different implementation provided in order to meet specific requirements.

The `Mail` object has the following public properties
* int $id - Uid of the email
* int $msgNo - Message number of the email
* string $subject - Subject line
* int $date - Date as timestamp
* array $to - List of address objects that email was sent to
* array $cc - List of address objects that email cc'd
* array $bcc - List of address objects that email bcc'd
* array $sender - List of address objects of email sender
* array $from - List of address objects email sent from
* array $replyTo - List of address objects of reply to
* bool $recent - Recent flag
* bool $unseen - Unseen flag
* bool $flagged - Flagged flag
* bool $answered - Answered flag
* bool $deleted - Deleted flag
* bool $draft - Draft flag

An address object (see `to`, `cc`, `bcc`, `sender`, `from`, `replyTo` properties above) is a standard object that has an `address` property, that contains the email address as `'example@gmail.com'`, a `name` property that contains the name corresponding to the email address as `'John Smith'`, and the domain of the address as `'gmail.com'`.

### getStream()
Get the underlying imap stream resource

### header($key, $default = null)
Find an email header value, or return $default
* string $key - The name of the header
* mixed $default - Return this value if no header called $key is found

### body($mimeType = 'html')
Get the parts of the email that correspond to the mimeType, and piece them together as a string.
* string $mimeType - The mime type the mail should be returned as. Defaults to 'html'

### fullBody()
Gets the full body structure of the email as an array. The parts of the email are flattened by the mail factory, and are accessible by key. The keys are the part numbers as 1, 1.1, 2, 2.1, 2.1.1 etc.

### move($mailbox)
Move the email to another mailbox
* string $mailbox - Name of the mailbox to move mail to

Returns boolean for successful move

# Contributing
Contributions are welcome. This follows the PSR-2 coding standard and the PSR-4 autoloading standard.


CHANGELOG.md
# Release Notes

## v1.0.0 (2016-09-07)

Initial Release


src/MailMap/MailFactory.php
<?php

namespace MailMap;

use InvalidArgumentException;
use MailMap\Contracts\MailFactory as FactoryContract;
use MailMap\Contracts\Mail as MailContract;
use MailMap\Mail;

class MailFactory implements FactoryContract
{
    /**
     * Default charset
     *
     * @var string
     */
    protected static $charset = 'UTF-8';

    /**
     * Flags to set when fetching headers/structure etc
     *
     * @var int Bit-mask
     */
    protected static $fetchFlags = FT_UID;

    /**
     * Flags to set when fetching parts of mail body
     *
     * @var int Bit-mask
     */
    protected static $bodyFlags = FT_UID | FT_PEEK;

    /**
     * The type of Mail wrapper this factory creates.
     *
     * Must be an implementation of \MailMap\Contracts\Mail
     *
     * @var string
     */
    private $mailClass;

    /**
     * Set the Mail wrapper class. Defaults to
     * the provided Mail implementation
     *
     * @param string $mailClass
     * @throws \InvalidArgumentException
     */
     public function __construct($mailClass = Mail::class)
     {
         if (!is_subclass_of($mailClass, MailContract::class)) {
             throw new InvalidArgumentException(sprintf('%s must implement %s', $mailClass, MailContract::class));
         }

         $this->mailClass = $mailClass;
     }

    /**
     * Create new mail from uid
     *
     * @param  int $uid
     * @param  resource $stream
     * @return \MailMap\Contracts\Mail
     */
    public function create($uid, $stream)
    {
        return new $this->mailClass(
            $uid,
            $stream,
            $this->parseMailHeader($stream, $uid),
            $this->parseHeaders($stream, $uid),
            $this->parseBody($stream, $uid)
        );
    }

    /**
     * Parse the mail header
     *
     * @param  resource $stream
     * @param  int $uid
     * @return array
     */
    protected function parseMailHeader($stream, $uid)
    {
        $header = imap_headerinfo($stream, imap_msgno($stream, $uid));

        return [
            'msgNo' => (int) $header->Msgno,
            'subject' => $this->parseSubject($header->subject),
            'date' => $header->udate,
            'to' => isset($header->to)
                ? $this->parseAddress($header->to) : null,
            'cc' => isset($header->cc)
                ? $this->parseAddress($header->cc) : null,
            'bcc' => isset($header->bcc)
                ? $this->parseAddress($header->bcc) : null,
            'sender' => isset($header->sender)
                ? $this->parseAddress($header->sender) : null,
            'from' => isset($header->from)
                ? $this->parseAddress($header->from) : null,
            'replyTo' => isset($header->reply_to)
                ? $this->parseAddress($header->reply_to) : null,
            'recent' => $this->parseFlag('R', $header->Recent),
            'unseen' => $this->parseFlag('U', $header->Unseen),
            'flagged' => $this->parseFlag('F', $header->Flagged),
            'answered' => $this->parseFlag('A', $header->Answered),
            'deleted' => $this->parseFlag('D', $header->Deleted),
            'draft' => $this->parseFlag('X', $header->Draft)
        ];
    }

    /**
     * Parse email addresses from header
     *
     * @param  array $addresses
     * @return array
     */
    protected function parseAddress(array $addresses)
    {
        return array_map(function ($address) {
            return (object) [
                'address' => $address->mailbox.'@'.$address->host,
                'domain' => $address->host,
                'name' => isset($address->personal)
                    ? static::convertBodyEncoding($address->personal)
                    : null
            ];
        }, array_filter($addresses, function ($address) {
            return property_exists($address, 'mailbox') && strtolower($address->mailbox) !== 'undisclosed-recipients';
        }));
    }

    /**
     * Check if given flag is set
     *
     * @param  string $flagChar
     * @param  string $flag
     * @return bool
     */
    protected function parseFlag($flagChar, $flag)
    {
        return strtoupper($flagChar) === strtoupper($flag);
    }

    /**
     * Parse the subject of the email
     *
     * @param  string $subject
     * @return string
     */
    protected function parseSubject($subject)
    {
        $subject = imap_mime_header_decode($subject);

        $subjectParts = array_map(function ($subj) {
            if (strtolower($subj->charset) === 'default') {
                return $subj->text;
            }
            return iconv($subj->charset, static::$charset, $subj->text);
        }, $subject);

        return implode('', $subjectParts);
    }

    /**
     * Parse the email headers into array structure
     *
     * @param  resource $stream
     * @param  int $uid
     * @return array
     */
    protected function parseHeaders($stream, $uid)
    {
        $rawHeaders = imap_fetchheader($stream, $uid, static::$fetchFlags);

        $headers = [];
        $key = null;

        foreach (explode("\n", $rawHeaders) as $line) {
            // Test if continutation
            if ($line !== '' && !preg_match('/^\s/', $line)) {
                list($key, $val) = explode(':', $line, 2);
                $headers[$key] = trim($val);
            } elseif (!is_null($key) && ($trimmed = trim($line)) !== '') {
                $headers[$key] = $headers[$key].$trimmed;
            }
        }

        foreach($headers as $headerKey => $headerValue) {
            if (preg_match('/(\w+)=(\S+);(?:\s|$)/', $headerValue)) {
                $headers[$headerKey] = $this->parseNestedKeyPairHeader($headerValue);
            }
        }

        return $headers;
    }

    /**
     * Parse the any nested key-pair values in the header value
     *
     * @param  string $nestedPairs
     * @return array
     */
    protected function parseNestedKeyPairHeader($nestedPairs)
    {
        return array_reduce(explode(';', $nestedPairs), function($result, $pair) {
            $keyVal = explode('=', $pair, 2);
            $key = trim(array_shift($keyVal));
            $val = trim(implode('=', $keyVal));
            if ($key !== '' && $val !== '') {
                $result[$key] = $val;
            }
            return $result;
        }, []);
    }

    /**
     * Parse the email body and its parts
     *
     * @param  resource $stream
     * @param  int $uid
     * @return array
     */
    protected function parseBody($stream, $uid)
    {
        $struct = imap_fetchstructure($stream, $uid, static::$fetchFlags);

        if (!isset($struct->parts)) {
            return array_filter([$this->parsePart($stream, $uid, $struct)]);
        }

        $parsedParts = [];
        foreach ($this->flattenParts($struct->parts) as $partNo => $part) {
            $parsedParts[] = $this->parsePart($stream, $uid, $part, $partNo);
        }
        return array_filter($parsedParts);
    }

    /**
     * Flatten email structure recursively.
     *
     * Parts re-keyed as 1, 1.1, 2, 2.1, 2.1.1 etc
     *
     * @param  array $parts
     * @param  array $flattened
     * @param  string $pre
     * @param  int $idx
     * @param  bool $full
     * @return array
     */
    protected function flattenParts(array $parts, array $flattened = [], $pre = '', $idx = 1, $full = true)
    {
        foreach ($parts as $part) {
            $key = $pre.$idx;
            $flattened[$key] = $part;
            if (isset($part->parts)) {
    			if ($part->type == TYPEMULTIPART) {
    				$flattened = $this->flattenParts($part->parts, $flattened, $key.'.', 1, false);
    			} elseif ($full) {
    				$flattened = $this->flattenParts($part->parts, $flattened, $key.'.');
    			} else {
    				$flattened = $this->flattenParts($part->parts, $flattened, $pre);
    			}
    			unset($flattened[$key]->parts);
    		}
    		$idx++;
        }

        return $flattened;
    }

    /**
     * Parse individual part of the structure
     *
     * @param  resource $stream
     * @param  int $uid
     * @param  stdClass $part
     * @param  int $partNo
     * @return stdClass | null
     */
    protected function parsePart($stream, $uid, $part, $partNo = 1)
    {
        if (!in_array($part->type, [TYPETEXT, TYPEMULTIPART])) {
            return null;
        }

        $body = imap_fetchbody($stream, $uid, $partNo, static::$bodyFlags);

        $body = static::decode($body, $part->encoding);
        if ('' !== ($charset = $this->findCharset($part))) {
            $body = static::convertBodyEncoding($body, $charset, $part->encoding);
        }

        return (object) [
            'body' => $body,
            'mime_type' => strtolower($part->subtype)
        ];
    }

    /**
     * Decode the body text
     *
     * @param  string $body
     * @param  int $encoding
     * @return string
     */
    protected static function decode($body, $encoding)
    {
        if (ENCQUOTEDPRINTABLE === $encoding) {
            return quoted_printable_decode($body);
        }

        if (ENCBASE64 === $encoding) {
            return base64_decode($body);
        }

        return $body;
    }

    /**
     * Convert body encoding from given coding to default
     *
     * @param  string $body
     * @param  string $charset
     * @param  int $encoding
     * @return string
     */
    protected static function convertBodyEncoding($body, $charset = '', $encoding = -1)
    {
        if ('' === $charset) {
            $charset = static::$charset;
        }

        if (!mb_check_encoding($body, $charset)) {
            $charset = mb_detect_encoding($body);
        }

        if ($charset === static::$charset) {
            return $body;
        }

        if (!in_array($charset, mb_list_encodings())) {
            $charset = $encoding === ENC7BIT ? 'US-ASCII' : static::$charset;
        }

        return mb_convert_encoding($body, static::$charset, $charset);
    }

    /**
     * Locate the charset within the struct
     *
     * @param  stdClass $struct
     * @return string
     */
    protected function findCharset($struct)
    {
        if (isset($struct->parameters)) {
            foreach ($struct->parameters as $param) {
                if (strtolower($param->attribute) === 'charset') {
                    return strtoupper($param->value);
                }
            }
        }
        if (isset($struct->dparameters)) {
            foreach ($struct->dparameters as $param) {
                if (strtolower($param->attribute) === 'charset') {
                    return strtoupper($param->value);
                }
            }
        }

        return '';
    }
}
